common/src/main/java/com/wynntils/features/ui/CustomPlayerListFeature.java

```diff
@@ -0,0 +1,242 @@
+                    .toList();
+
+            return lastPlayers;
+        }
+
+        @Override
+        public void render(PoseStack poseStack, MultiBufferSource bufferSource, float partialTicks, Window window) {
+            if (!McUtils.options().keyPlayerList.isDown() && animationProgress <= 0.0) return;
+            double animation = getAnimation();
+
+            renderPlayerList(poseStack, animation);
+        }
+
+        private void renderPlayerList(PoseStack poseStack, double animation) {
+            if (animation < 1) {
+                RenderUtils.enableScissor(
+                        (int) (getRenderX() + ROLL_WIDTH + HALF_WIDTH - HALF_WIDTH * animation),
+                        0,
+                        (int) (WIDTH * animation),
+                        McUtils.mc().getWindow().getScreenHeight());
+            }
+
+            renderBackground(poseStack);
+
+            float currentDist = getRenderX() + ROLL_WIDTH + 55;
+            float categoryStart = getRenderY() + 18;
+            renderCategoryTitle(poseStack, "Friends", currentDist, categoryStart);
+            currentDist += DISTANCE_BETWEEN_CATEGORIES;
+            renderCategoryTitle(poseStack, Models.WorldState.getCurrentWorldName(), currentDist, categoryStart);
+            currentDist += DISTANCE_BETWEEN_CATEGORIES;
+            renderCategoryTitle(poseStack, "Party", currentDist, categoryStart);
+            currentDist += DISTANCE_BETWEEN_CATEGORIES;
+            renderCategoryTitle(poseStack, "Guild", currentDist, categoryStart);
+
+            renderPlayerNames(poseStack, categoryStart);
+
+            if (animation < 1) {
+                RenderUtils.disableScissor();
+            }
+
+            float middle = getRenderX() + HALF_WIDTH + ROLL_WIDTH;
+            renderRoll(poseStack, (float) (middle - ROLL_WIDTH + 2 - HALF_WIDTH * animation));
+            renderRoll(poseStack, (float) (middle + HALF_WIDTH * animation));
+        }
+
+        @Override
+        public void renderPreview(
+                PoseStack poseStack, MultiBufferSource bufferSource, float partialTicks, Window window) {
+            renderPlayerList(poseStack, 1);
+        }
+
+        private void renderRoll(PoseStack poseStack, float middle) {
+            RenderUtils.drawTexturedRect(
+                    poseStack,
+                    Texture.PLAYER_INFO_OVERLAY.resource(),
+                    middle,
+                    getRenderY(),
+                    0,
+                    ROLL_WIDTH,
+                    Texture.PLAYER_INFO_OVERLAY.height(),
+                    0,
+                    0,
+                    ROLL_WIDTH,
+                    Texture.PLAYER_INFO_OVERLAY.height(),
+                    Texture.PLAYER_INFO_OVERLAY.width(),
+                    Texture.PLAYER_INFO_OVERLAY.height());
+        }
+
+        private void renderPlayerNames(PoseStack poseStack, float categoryStart) {
+            List<StyledText> players = getAvailablePlayers();
```

Comment:
Since this is a potentially expensive operation, and not strictly just "rendering", can you please move it up one level to renderPlayerList, and pass the `players` as an argument to it instead?